src/parser.ts

```diff
@@ -0,0 +1,90 @@
+import { NepaliMonthsData, DateFormats, DaysInMonthsMappingData } from './data'
+import { generateDateFormatOrder } from 'utils/generateDateFormatOrder'
+
+const MonthNames = NepaliMonthsData.map((month) => month.en)
+
+type ParseResult = {
+  year?: number
+  month?: number
+  day?: number
+}
+
+/**
+ * Parses a date string using predefined date formats.
+ *
+ * @param {string} dateString - The date string to be parsed.
+ * @returns {(ParseResult | string)} - The parsed date components or "Invalid Date" if parsing fails.
+ * @example parse('2077-01-01') // { year: 2077, month: 1, day: 1 }
+ */
+export const parse = (dateString: string): ParseResult | string => {
+  for (const dateFormat of DateFormats) {
+    const match = dateString.match(dateFormat.regex)
+    if (match) {
+      const parsedDate: ParseResult = {}
+      const order = generateDateFormatOrder(dateFormat.format)
+      order.forEach((component, index) => {
+        if (
+          component === 'month' &&
+          MonthNames.indexOf(match[index + 1]) >= 0
+        ) {
+          parsedDate[component] = MonthNames.indexOf(match[index + 1]) + 1
+        } else {
+          if (component === 'year' && match[index + 1].length < 4) {
+            if (match[index + 1].length === 2)
+              parsedDate[component] = parseInt('20' + match[index + 1])
+            else parsedDate[component] = parseInt('2' + match[index + 1])
+          } else {
+            parsedDate[component] = parseInt(match[index + 1])
+          }
+        }
+      })
+
+      // If month or day is not provided, default to 1
+      if (!parsedDate.month) {
+        parsedDate.month = 1
+      }
+
+      if (!parsedDate.day) {
+        parsedDate.day = 1
+      }
+
+      if (!isDayValid(parsedDate.year, parsedDate.month, parsedDate.day)) {
+        return 'Invalid Date'
+      }
+      return parsedDate
+    }
+  }
+  return 'Invalid Date'
+}
+
+/**
+ * Checks if a given day is valid for a given year and month.
+ *
+ * @param {number} year - The year.
+ * @param {number} month - The month (1-12).
+ * @param {number} day - The day.
+ * @returns {boolean} `true` if the day is valid, otherwise `false`.
+ */
+export const isDayValid = (year?: number, month?: number, day?: number) => {
+  if (year === undefined || month === undefined || day === undefined) {
+    return false
+  }
+  if (month != undefined && (month < 1 || month > 12)) {
```

Comment:
```suggestion
  if (month< 1 || month > 12) {
```
since we know month is not going to be undefined at this point